File: Binary Search/875. Koko Eating Bananas/solution.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def canFinish(speed):
            k = 0
            for p in piles:
                # k += ceil(p/speed)
                k += (p+(speed-1))//speed
            return k <= h

        l, r = 1, max(piles)
        while l < r:
            mid = l + (r-l)//2
            if canFinish(mid):
                r = mid
            else:
                l = mid+1
        return l

File: Binary Search/875. Koko Eating Bananas/solution.cpp
class Solution {
public:
    int check(vector<int>& piles, int h, int k) {
        for (auto& num : piles) {
            h -= (num+k-1)/k; // ceil(num/k)
        }
        return h >= 0;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int l=1, r=*max_element(piles.begin(), piles.end());

        while (l < r) {
            int mid = l + (r-l)/2;

            if (check(piles, h, mid)) {
                r = mid;
            } else {
                l = mid+1;
            }
        }
        return l;
    }
};